server/src/test/java/org/elasticsearch/health/HealthServiceTests.java

```diff
@@ -84,11 +132,78 @@ public void testMissingIndicator() {
                 createMockHealthIndicatorService(shardsAvailable)
             )
         );
+        NodeClient client = getTestClient(HealthInfo.EMPTY_HEALTH_INFO);
+        assertGetHealthThrowsException(
+            service,
+            client,
+            "indicator99",
+            false,
+            ResourceNotFoundException.class,
+            "Did not find indicator indicator99",
+            true
+        );
+    }
 
-        expectThrows(ResourceNotFoundException.class, "Did not find indicator indicator99", () -> service.getHealth("indicator99", false));
+    private <T extends Throwable> void assertGetHealthThrowsException(
+        HealthService service,
+        NodeClient client,
+        String indicatorName,
+        boolean explain,
+        Class<T> expectedType,
+        String expectedMessage,
+        boolean expectedInOnFail
+    ) throws Exception {
+        AtomicBoolean onFailureCalled = new AtomicBoolean(false);
+        ActionListener<List<HealthIndicatorResult>> listener = getExpectThrowsActionListener(
+            onFailureCalled,
+            expectedType,
+            expectedMessage
+        );
+        try {
+            service.getHealth(client, indicatorName, explain, listener);
+        } catch (Throwable t) {
+            if (expectedInOnFail || (expectedType.isInstance(t) == false)) {
+                throw new RuntimeException("Unexpected throwable", t);
+            } else {
+                // Expected
+                if (expectedMessage != null) {
+                    assertThat(t.getMessage(), equalTo(expectedMessage));
+                }
+            }
+        }
+        if (expectedInOnFail) {
+            assertBusy(() -> assertThat(onFailureCalled.get(), equalTo(true)));
+        }
     }
 
-    public void testPreflightIndicatorResultsPresent() {
+    private <T extends Throwable> ActionListener<List<HealthIndicatorResult>> getExpectThrowsActionListener(
+        AtomicBoolean onFailureCalled,
+        Class<T> expectedType,
+        String expectedMessage
+    ) {
+        return new ActionListener<>() {
+            @Override
+            public void onResponse(List<HealthIndicatorResult> healthIndicatorResults) {
+                fail("Unexpected");
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                if (expectedType.isInstance(e)) {
+                    if (expectedMessage == null) {
+                        onFailureCalled.set(true);
+                    } else {
+                        assertThat(e.getMessage(), equalTo(expectedMessage));
+                        onFailureCalled.set(true);
+                    }
+                } else {
+                    throw new RuntimeException("Unexpected exception", e);
```

Comment:
Should we use the `fail` assertion here too, like in `onResponse`?

Reply:
There's not a version of `fail` that takes an exception (and keeps track of the stack trace) is there? If we do run into a bug in the future here I thought it would be nice to have the stack trace to make it easier to track down the problem.

Reply:
Ah, I see, okay then this is better than fail.